flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java

```diff
@@ -75,10 +128,293 @@ public DefaultScheduler(
+
+		return false;
+	}
+
+	private void maybeHandleTaskFailure(final TaskExecutionState taskExecutionState, final ExecutionVertexID executionVertexId) {
+		if (taskExecutionState.getExecutionState() == ExecutionState.FAILED) {
+			final Throwable error = taskExecutionState.getError(userCodeLoader);
+			handleTaskFailure(executionVertexId, error);
+		}
+	}
+
+	private void handleTaskFailure(final ExecutionVertexID executionVertexId, final Throwable error) {
+		final FailureHandlingResult failureHandlingResult = executionFailureHandler.getFailureHandlingResult(executionVertexId, error);
+		maybeRestartTasks(failureHandlingResult);
+	}
+
+	private void maybeRestartTasks(final FailureHandlingResult failureHandlingResult) {
+		if (failureHandlingResult.canRestart()) {
+			restartTasksWithDelay(failureHandlingResult);
+		} else {
+			failJob(failureHandlingResult.getError());
+		}
+	}
+
+	private void restartTasksWithDelay(final FailureHandlingResult failureHandlingResult) {
+		final Set<ExecutionVertexID> verticesToRestart = failureHandlingResult.getVerticesToRestart();
+
+		final Set<ExecutionVertexVersion> executionVertexVersions =
+			new HashSet<>(executionVertexVersioner.recordVertexModifications(verticesToRestart).values());
+
+		final CompletableFuture<?> cancelFuture = cancelTasksAsync(verticesToRestart);
+
+		delayExecutor.schedule(
+			() -> FutureUtils.assertNoException(
+				cancelFuture.handleAsync(restartTasksOrHandleError(executionVertexVersions), getMainThreadExecutor())),
+			failureHandlingResult.getRestartDelayMS(),
+			TimeUnit.MILLISECONDS);
+	}
+
+	private BiFunction<Object, Throwable, Void> restartTasksOrHandleError(final Set<ExecutionVertexVersion> executionVertexVersions) {
+		return (Object ignored, Throwable throwable) -> {
+
+			if (throwable == null) {
+				final Set<ExecutionVertexID> verticesToRestart = executionVertexVersioner.getUnmodifiedExecutionVertices(executionVertexVersions);
+				schedulingStrategy.restartTasks(verticesToRestart);
+			} else {
+				failJob(throwable);
+			}
+			return null;
+		};
+	}
+
+	private CompletableFuture<?> cancelTasksAsync(final Set<ExecutionVertexID> verticesToRestart) {
+		final List<CompletableFuture<?>> cancelFutures = verticesToRestart.stream()
+			.map(this::cancelExecutionVertex)
+			.collect(Collectors.toList());
+
+		return FutureUtils.combineAll(cancelFutures);
+	}
+
+	private CompletableFuture<?> cancelExecutionVertex(final ExecutionVertexID executionVertexId) {
+		return executionVertexOperations.cancel(getExecutionVertex(executionVertexId));
 	}
 
 	@Override
-	public void startScheduling() {
-		throw new UnsupportedOperationException();
+	public void scheduleOrUpdateConsumers(final ResultPartitionID partitionId) {
+		final Optional<ExecutionVertexID> producerVertexId = getExecutionVertexId(partitionId.getProducerId());
+		if (producerVertexId.isPresent()) {
+			updateConsumers(partitionId);
+			schedulingStrategy.onPartitionConsumable(producerVertexId.get(), partitionId);
```

Comment:
A partition is consumable in 2 cases:
1. a pipelined partition has data produced (including notified finished from TM side)
2. a blocking partition is finished and all other partitions in the its parent IntermediateResult is finished

Here we only cover the first case, though it is not problematic at the moment since the `LazyFromSourcesSchedulingStrategy` is doing scheduling based on the FINISHED task state rather than consumable partition state. 
Future scheduling strategies, however, may find the needed state updates of partitions are missing.

Besides, I think the philosophy of `lazy_from_sources` scheduling is to schedule a task when its inputs are ready, naming the needed result partitions are consumable. It actually does not directly care whether a task is finished, though a finished task will lead its partitions to be consumable.

The right place to catch the partition state state changes is in `Execution#scheduleOrUpdateConsumers` currently. 

I think we should reconsider how we pass the partition state updates and how to better implement `LazyFromSourcesSchedulingStrategy`. And sorry for not recognizing this potential issue in 1.9.

It can be follow up work after we have made `DefaultScheduler` in a working state.

Reply:
I don't fully understand your comment. Can you explain how you want to improve `LazyFromSourcesSchedulingStrategy`?

> It actually does not directly care whether a task is finished, though a finished task will lead its partitions to be consumable.

That is true but aren't we intentionally not sending the `scheduleOrUpdateConsumers` message to avoid RPC overhead? Whether or not a producer sends these messages is controlled via `DeploymentOption`.

Reply:
I mean to make `LazyFromSourcesSchedulingStrategy` trigger scheduling in `onPartitionConsumable` instead of in `onExecutionStateChange`.
This also requires the scheduler to invoke `onPartitionConsumable` on a vertex when its input partition is consumable due to the producer vertex finishes.
In this way, the `LazyFromSourcesSchedulingStrategy` only needs to react to the partition consumable state changes.

> That is true but aren't we intentionally not sending the scheduleOrUpdateConsumers message to avoid RPC overhead? Whether or not a producer sends these messages is controlled via DeploymentOption.

I did not mean to change the `scheduleOrUpdateConsumers` message sending. But that  `SchedulingStrategy#onPartitionConsumable` should be notified both when a `scheduleOrUpdateConsumers` message comes, **and a finished vertex also finishes its produced partitions**.
We can skip the notification only when a partition is pipelined and it disables the `sendScheduleOrUpdateConsumerMessage` in `DeploymentOption`.

Reply:
Ok, I understand. I can see that it would be more consistent to make scheduling decisions only based on `SchedulingStrategy#onPartitionConsumable()` in `LazyFromSourcesSchedulingStrategy`. At the moment we (re-)implemented partition tracking inside `LazyFromSourcesSchedulingStrategy`. It could be generalized and moved outside of the strategy. I am not sure if the `ExecutionGraph` is the right place to do it, however.

Reply:
> It can be follow up work after we have made DefaultScheduler in a working state.

Yes.

Reply:
I thin this is partition state management and will eventually be part of the scheduler, similar to what we may do to vertex state management.

Currently `ExecutionGraph` is already doing this. `Execution#scheduleOrUpdateConsumers` is invoked when and only when a partition becomes consumable. Before factored out, I think it's the best place to notify the scheduler about partition consumable.

Reply:
[FLINK-14234](https://issues.apache.org/jira/browse/FLINK-14234) is opened to track this.